opentelemetry/src/trace/recruited_tracer.rs

```diff
@@ -0,0 +1,57 @@
+//! Abstracts away details for acquiring a `Tracer` by instrumented libraries.
+use crate::global::BoxedTracer;
+use once_cell::sync::OnceCell;
+use std::fmt::Debug;
+
+/// Holds either a borrowed `BoxedTracer` or a factory that can produce one when
+/// and if needed.
+///
+/// This unifies handling of obtaining a `Tracer` by library code optimizing for
+/// common cases when it will never be needed.
+#[derive(Debug)]
+pub struct RecruitedTracer<'a> {
+    variant: Variant<'a>,
+    tracer: OnceCell<BoxedTracer>,
+}
+
+enum Variant<'a> {
+    Borrowed(&'a BoxedTracer),
+    Lazy(&'a dyn Fn() -> BoxedTracer),
+}
+use Variant::*;
```

Comment:
I prefer not to import enum variants at the module level, IMO this should be restricted to function-level scopes.